File: bigData/hw5/MATLAB/mapping.m
function Xmap = mapping(X,degree)
[m,n] = size(X); % m is number of variables
power = selectk(0:degree,n);
ind = (sum(power,2)<=degree);
power = power(ind,:);
p = size(power,1); % number of  polynomial terms
Xmap = ones(m,1);
for i = 1:p
    aterm = ones(m,1);
    for j = 1:n
        aterm = aterm .* X(:,j).^power(i,j);    
    end
    Xmap(:,end+1) = aterm;
end

File: bigData/hw5/R/mapping.R
source('selectk.R')
mapping <- function(X, degree){
    m <- nrow(X)
    n <- ncol(X)
    power <- selectk(0:degree, n)
    ind <- (rowSums(power) <= degree)
    power <- power[ind, ]
    p <- nrow(power)
    Xmap <- matrix(0, m, p)
    for (i in 1:p){
        aterm <- rep(1, m)
        for (j in 1:n){
            aterm <- aterm * X[, j]^power[i, j]
        }
        Xmap[, i]  <- aterm
    }
    return(Xmap)
}
